src/hardware/vga_draw.cpp

```diff
@@ -1198,6 +1199,39 @@ static void maybe_aspect_correct_tall_modes(double &current_ratio)
 	current_ratio *= 2;
 }
 
+// Pre-VGA color modes often ran with a horizonal sync rate of 15.7 kHz
+// (200-lines) or 21.85 KHz (350-lines), both producing a vertical sync rate of
+// 60 Hz on those old systems, however newer VGA systems often run at a slightly
+// higher clock rate. So we adjust the clock when not running a pre-VGA system.
+// Ref: http://minuszerodegrees.net/video/bios_video_modes.htm
+static uint32_t get_clock_bias_for_pre_vga_color_mode(const uint32_t clock_hz,
+                                                      const uint32_t htotal,
+                                                      const uint32_t vtotal)
+{
+	constexpr uint32_t pre_vga_color_modes = (M_EGA | M_CGA2 |
+	                                          M_CGA2_COMPOSITE | M_CGA4 |
+	                                          M_CGA4_COMPOSITE | M_CGA16 |
+	                                          M_TANDY2 | M_TANDY4 | M_TANDY16);
+
+	const auto is_pre_vga_color_mode = bit::any(static_cast<uint32_t>(vga.mode),
+	                                            pre_vga_color_modes);
+
+	const auto is_200_of_350_line_mode = (CurMode->sheight == 200 ||
+	                                      CurMode->sheight == 350);
+
+	if (!is_pre_vga_color_mode || !is_200_of_350_line_mode || !is_machine(MCH_VGA))
+		return 0;
+
+	constexpr auto pre_vga_color_200_350_vsync_rate_hz = 60u;
+
+	const auto target_clock_hz = pre_vga_color_200_350_vsync_rate_hz *
+	                             htotal * vtotal;
+
+	assert(clock_hz >= target_clock_hz);
+	const auto bias_hz = clock_hz - target_clock_hz;
+	return bias_hz;
```

Comment:
I'm slightly confused about this bias calculation. Isn't this just simply a matter of setting the refresh rate for low-res CGA/EGA-like modes to 60Hz on non-VGA, and 70Hz on VGA? So not quite sure why the bias calculation is needed, but I'm not familiar with this part of the code.

Reply:
Also related:

- MDA refresh rate is 50Hz (probably doesn't matter much if we don't emulate that because it's text-only anyway)
- Old VGA cards (before SVGA and VESA) only supported 60Hz in 640x480 / 16 color mode. Not sure if that's handled properly, so maybe there needs to be a special case for it as well? (And I actually don't know what a later SVGA/VESA card does in that mode; probably they just use 60Hz for backwards compatibility).

Reply:
Some context:

`VGA_SetupDrawing(..)` is a huge workhorse function that emulates  function this takes the game's desired mode via the set crtc registers and the card's clock rate and produces final sizes and timings used to setup the mode.

Here's a small chunks, reading those CRTC registers:
``` c++
htotal |= (vga.s3.ex_hor_overflow & 0x1) << 8;
htotal += 3;
hdend |= (vga.s3.ex_hor_overflow & 0x2) << 7;
hbend |= (vga.crtc.end_horizontal_retrace&0x80) >> 2;
hbstart |= (vga.s3.ex_hor_overflow & 0x4) << 6;
hrstart |= (vga.s3.ex_hor_overflow & 0x10) << 4;			
vtotal |= (vga.crtc.overflow & 0x20) << 4;
vtotal |= (vga.s3.ex_ver_overflow & 0x1) << 10;
vdend |= (vga.crtc.overflow & 0x40) << 3; 
vdend |= (vga.s3.ex_ver_overflow & 0x2) << 9;
vbstart |= (vga.crtc.maximum_scan_line & 0x20) << 4;
vbstart |= (vga.s3.ex_ver_overflow & 0x4) << 8;
vrstart |= ((vga.crtc.overflow & 0x80) << 2);
vrstart |= (vga.s3.ex_ver_overflow & 0x10) << 6;
vbend = vga.crtc.end_vertical_blanking & 0x7f;
```

For VGA cards, the clock rate is card-specific (Tseng, S3, Paradise have different clock generator functions), while the non-VGA cards only supported a couple clocks, so those are just defined as-is:

```c++
if (svga.get_clock) {
	clock = svga.get_clock();
} else {
	switch ((vga.misc_output >> 2) & 3) {
	case 0:	
		clock = (machine==MCH_EGA) ? 14318180 : 25175000;
		break;
	case 1:
	default:
		clock = (machine==MCH_EGA) ? 16257000 : 28322000;
		break;
	}
}

// Adjust the VGA clock frequency based on the Clocking Mode Register's
// 9/8 Dot Mode. See Timing Model: https://wiki.osdev.org/VGA_Hardware
clock /= (vga.seq.clocking_mode & 1) ? 8 : 9;
```

The clock can be further adjusted based on clocking-mode. 

These sizes and timings combined with with card's clock rate to form the resulting video mode's final properties:

```
// Start and End of horizontal blanking
vga.draw.delay.hblkstart = static_cast<double>(hbstart) * 1000.0 / f_clock; //  milliseconds
vga.draw.delay.hblkend = static_cast<double>(hbend) * 1000.0 / f_clock;
// Start and End of horizontal retrace
vga.draw.delay.hrstart = static_cast<double>(hrstart) * 1000.0 / f_clock;
vga.draw.delay.hrend = static_cast<double>(hrend) * 1000.0 / f_clock;
// Start and End of vertical blanking
vga.draw.delay.vblkstart = static_cast<double>(vbstart) * vga.draw.delay.htotal;
vga.draw.delay.vblkend = static_cast<double>(vbend) * vga.draw.delay.htotal;
// Start and End of vertical retrace pulse
vga.draw.delay.vrstart = static_cast<double>(vrstart) * vga.draw.delay.htotal;
vga.draw.delay.vrend = static_cast<double>(vrend) * vga.draw.delay.htotal;
```

These timings and sizes "drive" the mode via DOSBox's emulated video timer functions, IRQ functions, and latch functions (and quite a view more). 

---

> Isn't this just simply a matter of setting the refresh rate for low-res CGA/EGA-like modes to 60Hz on non-VGA, and 70Hz on VGA

When `machine` is set to `ega`, `cga`, or `tandy`, then the int10 mode tables will use `ModeList_EGA` or `ModeList_OTHER` from the "int10_modes.cpp", which use "period-correct" mode paramaters and produce 60 Hz FPS per those original old cards -- so we're all good there.

The problem is when `machine = svga_s3 / paradise / tseng / ...` /and/ when an old CGA/EGA-mode is used.  These SVGA cards used slightly different mode parameters for the EGA and CGA modes (you can compare them in ModeList_VGA versus the EGA table), likely to try to be more compatible with these cards higher clock rates. So these old modes are driven at driven up around the 69 to 71 FPS range when using an SVGA card.

So the idea with this PR is that we slow the SVGA clock "by a little bit" to make it more period-correct, however we still leave all mode properties intact.

I opted to use a similar clock-adjustment approach, similar to this:

``` C++
// Adjust the VGA clock frequency based on the Clocking Mode Register's
// 9/8 Dot Mode. See Timing Model: https://wiki.osdev.org/VGA_Hardware
clock /= (vga.seq.clocking_mode & 1) ? 8 : 9;
```

Where we apply a change to the clock.  It makes it a bit more portable (the result from function could be moved elsewhere, and applied as a delta later on).

Reply:
The MDA modes get 50 Hz and the higher resolution modes already get 60 Hz - so no adjustment is needed.

Reply:
From a functional perspective, DOSBox's 'all purpose' VGA emulation is amazingly compatible, if you compare it versus the table here:

https://gona.mactar.hu/DOS_TESTS/

Most cards fails in multiple old game tests - where as DOSBox runs them faithfully.

Reply:
Adjusted the function to update the clock in-place.